drivers/photo-electronic/adpd188/iio_adpd188.c

```diff
@@ -0,0 +1,522 @@
+ * @param mask - Mask of the enabled channels.
+ * @return SUCCESS in case of success, error code otherwise.
+ */
+static int32_t iio_adpd188_prepare_data_read(void *dev, uint32_t mask)
+{
+	struct adpd188_iio_desc *iio_desc = (struct adpd188_iio_desc *)dev;
+	struct adpd188_dev *desc = iio_desc->drv_dev;
+
+	iio_desc->ch_mask = mask;
+
+	return adpd188_iio_normal_mode(desc);
+}
+
+/**
+ * @brief Disable all channels after buffer read.
+ * @param dev - Pointer to the IIO driver structure.
+ * @return SUCCESS in case of success, error code otherwise.
+ */
+static int32_t iio_adpd188_end_data_read(void *dev)
+{
+	struct adpd188_iio_desc *iio_desc = (struct adpd188_iio_desc *)dev;
+	struct adpd188_dev *desc = iio_desc->drv_dev;
+
+	iio_desc->ch_mask = 0;
+
+	return adpd188_iio_standby_mode(desc);
+}
+
+/**
+ * @brief Get index of a bit within a channel mask.
+ * @param [in] ch_mask - Channel mask.
+ * @param [in] last_idx - Last index found to be set.
+ * @param [out] new_index - Next bit that is set or -1 in case there is none.
+ * @return true in case a set bit is found, flase otherwise.
+ */
+static bool get_next_ch_idx(uint32_t ch_mask, uint32_t last_idx,
+			    uint32_t *new_idx)
+{
+	last_idx++;
+	if (last_idx >= 32)
+		return 0;
+	ch_mask >>= last_idx;
+	if (!ch_mask) {
+		*new_idx = -1;
+		return 0;
+	}
+	while (!(ch_mask & 1)) {
+		last_idx++;
+		ch_mask >>= 1;
+	}
+	*new_idx = last_idx;
+
+	return 1;
+}
+
+/**
+ * @brief Read consecutive samples in a buffer.
+ * @param dev - Pointer to the IIO driver structure.
+ * @param buff - Output buffer.
+ * @param nb_samples - Number of samples to take.
+ * @return SUCCESS in case of success, error code otherwise.
+ */
+static int32_t iio_adpd188_read_samples(void *dev, int32_t *buff,
+					uint32_t nb_samples)
+{
+	struct adpd188_iio_desc *iio_desc = (struct adpd188_iio_desc *)dev;
+	struct adpd188_dev *desc = iio_desc->drv_dev;
+	uint8_t byte_no;
+	uint32_t i, k, new_idx, last_idx, buff_idx = 0;
+	uint16_t data_buff[16];
+	int32_t ret;
+
+	for (i = 0; i < nb_samples; i++) {
+		do {
+			ret = adpd188_fifo_status_get(desc, &byte_no);
+			if (IS_ERR_VALUE(ret))
+				return FAILURE;
+		} while (byte_no < 32);
+
+		for (k = 0; k < 16; k++) {
+			ret = adpd188_reg_read(desc, ADPD188_REG_FIFO_ACCESS,
+					       (data_buff + k));
+			if (IS_ERR_VALUE(ret))
+				return FAILURE;
+		}
+		last_idx = -1;
+		while (get_next_ch_idx(iio_desc->ch_mask, last_idx,
+				       &new_idx)) {
+			buff[buff_idx] = data_buff[new_idx * 2];
+			buff[buff_idx++] = data_buff[new_idx * 2 + 1];
+			last_idx = new_idx;
+		}
+	}
+
+	return SUCCESS;
+}
+
+/**
+ * @brief Initialize the IIO driver and allocate memory.
+ * @param device - Pointer to the IIO driver handler structure.
+ * @param init_param - Pointer to the IIO driver initialization structure.
+ * @return SUCCESS in case of success, error code otherwise.
+ */
+int32_t adpd188_iio_init(struct adpd188_iio_desc **device,
```

Comment:
This is indeed application specific code. For this reason I would not put the mask inside the driver, because number of channels and what the channels mean varies in this device. A mask would not really be appropriate there, in my opinion.

Reply:
In my opinion then, I would add an enum adpd188_configurations in the adpd188_iio_init_param. And a function similar to https://github.com/analogdevicesinc/no-OS/blob/67477349aba606222d9a11c5d142fe19f9c6d77f/drivers/axi_core/iio_axi_adc/iio_axi_adc.h#L102
And in the adpd188_iio_init you configured the needed channels and a iio_descriptor with the selected configuration is returned.
I think this way the user will now how to configure new iio_channels if he want other configurations and will be useful if in the future we will extend this file.